tempesta_fw/http.c

```diff
@@ -660,24 +673,29 @@ tfw_http_req_zap_error(struct list_head *equeue)
 
 	list_for_each_entry_safe(req, tmp, equeue, fwd_list) {
 		list_del_init(&req->fwd_list);
-		switch(req->httperr.status) {
-		case 404:
-			tfw_http_send_404(req, req->httperr.reason);
-			break;
-		case 500:
-			tfw_http_send_500(req, req->httperr.reason);
-			break;
-		case 502:
-			tfw_http_send_502(req, req->httperr.reason);
-			break;
-		case 504:
-			tfw_http_send_504(req, req->httperr.reason);
-			break;
-		default:
-			TFW_WARN("Unexpected response error code: [%d]\n",
-				 req->httperr.status);
-			tfw_http_send_500(req, req->httperr.reason);
-			break;
+		if (!ret) {
+			switch(req->httperr.status) {
+			case 404:
+				ret = tfw_http_send_404(req, req->httperr.reason);
+				break;
+			case 500:
+				ret = tfw_http_send_500(req, req->httperr.reason);
+				break;
+			case 502:
+				ret = tfw_http_send_502(req, req->httperr.reason);
+				break;
+			case 504:
+				ret = tfw_http_send_504(req, req->httperr.reason);
+				break;
+			default:
+				TFW_WARN("Unexpected response error code: [%d]\n",
+					 req->httperr.status);
+				ret = tfw_http_send_500(req, req->httperr.reason);
+				break;
+			}
+		} else {
+			ss_close_sync(req->conn->sk, true);
```

Comment:
I am not sure I undestand. `tfw_http_send_XXX()` functions ultimately call `tfw_http_send_resp()` which handles errors. It appears that with this code you're doing the same job twice.

And yes, like noted by @ikoveshnikov, you actually close a wrong client connection - from the next request, if there's any. And that request may be from a different connection.

So, which place to handle errors is ultimately better? If it's inside `tfw_http_send_resp()`, then it looks like `tfw_http_send_XXX()` functions don't need to return an error code at all. Or, perhaps, cases like with HTTP response codes 302 and 304 may prompt handling these errors outside of `tfw_http_send_resp()`.

It appears that more analysis is needed for a solution that handles all cases in an elegant way.

Reply:
I inserted error handling logic into `tfw_http_req_zap_error` - from the considerations that if we faced with memory allocation error, it is better not to try sending remain error responses and just to close all those connections - to speed up memory reclamation. But now I see that this is wrong way: closing the connection immediately without trying to send the client an error message and possibly to leave connection alive - is not justified.
So, I removed error handling logic from `tfw_http_req_zap_error`.
I left the error code returning from the functions `tfw_http_send_*` since they are used in unit tests:
```
        EXPECT_EQ(tfw_http_send_502(mock.req, "sticky calculation"), 0);
```
But maybe in case of unit tests - it is worth to replace functions `tfw_http_send_*` with helper functions which return values.

Reply:
@keshonok thank you for review.